File: src/infrastructure/config/config-validation.ts
import { z } from 'zod'
import type { ReasoningLevel } from '../ai/ai.models'

export type PromptValidator = (value: string | undefined) => string | undefined

export const providerSchema = z.enum(['codex', 'opencode', 'openai-compatible'])
export const reasoningLevelSchema = z.enum([
  'provider-default',
  'none',
  'minimal',
  'low',
  'medium',
  'high',
  'xhigh'
])
export const reasoningLevelPromptSchema = reasoningLevelSchema

const emailSchema = z.string().trim().email('Informe um email válido.')
export const optionalEmailPromptSchema = z
  .string()
  .trim()
  .refine(
    (value) => value === '' || emailSchema.safeParse(value).success,
    'Informe um email válido ou deixe vazio.'
  )

function firstIssueMessage(error: { issues: ReadonlyArray<{ message: string }> }): string | undefined {
  return error.issues.at(0)?.message
}

export const validateOptionalEmail: PromptValidator = (value) => {
  const parsed = optionalEmailPromptSchema.safeParse(value ?? '')
  return parsed.success
    ? undefined
    : (firstIssueMessage(parsed.error) ?? 'Informe um email válido.')
}

export function parseReasoningLevel(value: unknown, fallback: ReasoningLevel): ReasoningLevel {
  if (value === undefined || value === null) return fallback
  if (typeof value === 'string' && value.trim() === '') return fallback
  const parsed = reasoningLevelSchema.safeParse(value)
  if (!parsed.success) {
    throw new Error(
      `Nível de thinking inválido (${typeof value}). Use provider-default, none, minimal, low, medium, high ou xhigh.`
    )
  }
  return parsed.data
}

export function parseProvider(value: string) {
  const parsed = providerSchema.safeParse(value)
  if (parsed.success) return parsed.data
  throw new Error(`Provider inválido: ${value}. Use codex, opencode ou openai-compatible.`)
}

File: apps/dart/lib/src/application/config/config_validation.dart
import 'config_models.dart';

const providerValues = <ProviderName>{'codex', 'opencode', 'openai-compatible'};

const reasoningLevelValues = <ReasoningLevel>{
  'provider-default',
  'none',
  'minimal',
  'low',
  'medium',
  'high',
  'xhigh',
};

ProviderName parseProvider(String value) {
  if (providerValues.contains(value)) return value;
  throw FormatException(
    'Provider inválido: $value. Use codex, opencode ou openai-compatible.',
  );
}

List<ProviderName> parseProviderList(String value) {
  final providers = value
      .split(',')
      .map((item) => item.trim())
      .where((item) => item.isNotEmpty)
      .map(parseProvider)
      .toList();
  if (providers.isEmpty) {
    throw const FormatException('Configure pelo menos um provider.');
  }
  return providers;
}

ReasoningLevel parseReasoningLevel(Object? value, ReasoningLevel fallback) {
  if (value == null) return fallback;
  if (value is String && value.trim().isEmpty) return fallback;
  if (value is String && reasoningLevelValues.contains(value)) return value;
  throw FormatException(
    'Nível de thinking inválido (${value.runtimeType}). Use provider-default, '
    'none, minimal, low, medium, high ou xhigh.',
  );
}

String? validateOptionalEmail(String? value) {
  final normalized = (value ?? '').trim();
  if (normalized.isEmpty) return null;
  final valid = RegExp(r'^[^\s@]+@[^\s@]+\.[^\s@]+$').hasMatch(normalized);
  return valid ? null : 'Informe um email válido ou deixe vazio.';
}
